File: frontend/src/api/structuring.ts
// ================================
// Text Structuring API Types
// ================================

import type { ProcessingStatus } from './ocr'

// POST /api/evaluations/:evaluationId/structuring/start - Start text structuring
export interface StartStructuringRequest {
    sheetIds?: string[] // Optional: specific sheets, otherwise all OCR-completed sheets
}

export interface StartStructuringResponse {
    success: boolean
    jobId: string
    message: string
    totalSheets: number
}

// GET /api/evaluations/:evaluationId/structuring/status - Get structuring status
export interface StructuringStatusResponse {
    success: boolean
    status: ProcessingStatus
    progress: {
        total: number
        completed: number
        failed: number
        percentage: number
    }
    startedAt?: string
    completedAt?: string
    error?: string
}

// GET /api/evaluations/:evaluationId/structuring/result/:sheetId - Get structured text
export interface StructuringResultResponse {
    success: boolean
    sheetId: string
    status: ProcessingStatus
    result?: StructuredText
    error?: string
}

export interface StructuredText {
    questions: StructuredQuestion[]
    metadata: {
        totalQuestions: number
        totalSteps: number
        averageComplexity: string
    }
}

export interface StructuredQuestion {
    questionNumber: number
    questionText: string
    answer: {
        rawText: string
        structuredSteps: AnswerStep[]
        identifiedAlgorithm?: string
        complexity?: string // e.g., "O(n log n)"
        keywords: string[]
        codeBlocks?: CodeBlock[]
    }
}

export interface AnswerStep {
    stepNumber: number
    description: string
    type: 'explanation' | 'code' | 'diagram' | 'calculation'
    content: string
}

export interface CodeBlock {
    language: string
    code: string
    lineNumbers: { start: number; end: number }
}

File: backend/app/schemas/structuring.py
"""
Text structuring schemas.
Matches the frontend contract in api/structuring.ts.
"""
from pydantic import BaseModel
from typing import Literal, Optional

from app.schemas.common import ProcessingStatus


# --- Answer components ---

class AnswerStep(BaseModel):
    step_number: int
    description: str
    type: Literal["explanation", "code", "diagram", "calculation"]
    content: str


class CodeBlock(BaseModel):
    language: str = "python"
    code: str
    line_numbers: dict  # {"start": int, "end": int}


class StructuredAnswer(BaseModel):
    raw_text: str
    structured_steps: list[AnswerStep]
    identified_algorithm: str | None = None
    complexity: str | None = None
    keywords: list[str] = []
    code_blocks: list[CodeBlock] = []


class StructuredQuestion(BaseModel):
    question_number: int
    question_text: str
    answer: StructuredAnswer


class StructuredTextMetadata(BaseModel):
    total_questions: int
    total_steps: int
    average_complexity: str = "N/A"


class StructuredText(BaseModel):
    questions: list[StructuredQuestion]
    metadata: StructuredTextMetadata


# --- API response models ---

class StructuringStartResponse(BaseModel):
    success: bool = True
    job_id: str
    message: str
    total_sheets: int


class StructuringProgress(BaseModel):
    total: int
    completed: int
    failed: int
    percentage: int


class StructuringStatusResponse(BaseModel):
    success: bool = True
    status: ProcessingStatus
    progress: StructuringProgress
    started_at: str | None = None
    completed_at: str | None = None
    error: str | None = None


class StructuringResultResponse(BaseModel):
    success: bool = True
    sheet_id: str
    status: ProcessingStatus
    result: StructuredText | None = None
    error: str | None = None
